aGHQ.qmd

```diff
@@ -194,13 +194,26 @@ d(y_i, \eta_i)&=-2\left[y_i\log(\mu_i)+(1-y_i)\log(1-\mu_i)\right]\\
 \quad i=1,\dots,n
 $$
 
-A Julia function to evaluate both the mean and the unit deviance can be written as
+We will store the response vector, $\bby$, the linear predictor, $\bbeta$, the vector of means, $\bbmu$, and the unit deviances as columns in a table.
+Because all the columns are vectors of the same type, `Float64`, we use a `MatrixTable`, which consists of a `Matrix` and a vector of column names, stored as `Symbol`s.
+As with other Table types, columns are accessed as properties and often within functions we use "destructuring"
+to access them by name.
+
+The `updatetbl!` method iterates over the rows updating the $\bbmu$ and `dev` entries from the `y` and $\bbeta$ values in that row.
 
 ```{julia}
 #| output: false
-function meanunitdev(y::T, η::T) where {T<:AbstractFloat}
-  expmη = exp(-η)
-  return (; μ=inv(1 + expmη), dev=2 * ((1 - y) * η + log1p(expmη)))
+function updatetbl!(
+  tbl::MatrixTable{Matrix{T}},
+) where {T<:AbstractFloat}
+  (; y, η, μ, dev) = tbl   # tbl must have columns with these names
+  @inbounds for i in eachindex(y, η, μ, dev)
+    ηi = η[i]
+    expmη = exp(-ηi)
+    μ[i] = inv(one(T) + expmη)
+    dev[i] = 2 * ((one(T) - y[i]) * ηi + log1p(expmη))
+  end
```

Comment:
because of the use of concrete eltypes and equal length vectors, each row is a constant distance away in memory, which means that the CPU prefetch may still work quite well -- you grab adjacent chunks of memory and load them into the L1/L2 cache and then you're grabbing elements out of that cache.